### ISSUE TITLE
Any possibility of doing an Appimage?

### ISSUE THREAD

**Issue by nathan2423** [cboxdoerfer/fsearch#85]
For those of us whose distributions don't have gtk+3 3.16 (see the thread on QT interface) would there be any possibility of providing an Appimage?

**Comment by cboxdoerfer** [comment]
I haven't looked into Appimages (or Flatpak and Snap packages) so far, but if it works well and doesn't require too much time it certainly might be a good idea - especially for the upcoming 0.1 release.

Btw. I'm also trying to lower the GTK+ 3 requirements. I just got rid of the usage of the gtk_widget_get_action_group API which was only available in 3.16. It wasn't really needed at all.

**Comment by probonopd** [comment]
Yes, it's best practice to develop against the oldest still-supported LTS/Enterprise distributions.

Can you add `trusty` to https://launchpad.net/%7Echristian-boxdoerfer/+archive/ubuntu/fsearch-daily? Then I could convert it to an AppImage relatively easily.

**Comment by cboxdoerfer** [comment]
@probonopd, I recently added support for Debian 8, but currently I don't have that much time for coding and there are other things I find more important (e.g. an improved database format and query syntax I've been working on the last couple of months).

But if anyone is willing to work on 16.04 support, it should be quite straight forward and I'll gladly accept contributions.

**Comment by nazar2sfive** [comment]
+1 for flatpak image please.

**Comment by probonopd** [comment]
This issue is about AppImage. Flatpak does not give you "1 app = 1 file" that you can easily move around.

**Comment by ghcatt** [comment]
second that. how about the chances for an appimage build in the future?

**Comment by jawz101** [comment]
I'd prefer snap or flatpak vs appimage:

>it's in a universal repo any distro can use (rather than search for an appimage on some website)

and snap because it auto updates.  I don't know if appimage or flatpak auto update but if the development is quite frequent it'd be nice to simply get the latest bug fixes

**Comment by probonopd** [comment]
It's not an either-or. Why not have both. I prefer to update only the apps I want to update, when I decide that I need an update. And keep the old version in parallel in case I need it.

**Comment by cboxdoerfer** [comment]
Yes, Flatpak builds are going to happen either way. I already have them working locally and with the 0.1 release they'll be hopefully available on flathub.org.

I briefly skimmed over the AppImage documentation and it seemed more complex than the Flatpak packaging, which took just a couple of minutes. So I'm not sure when and if AppImages are going to happen, but it sounds nice to integrated them with the Travis CI for up-to-date dev builds for all platforms.

**Comment by Philipp91** [comment]
While I had no trouble installing the PPA, I now get the daily updates in the form of notifications from my update manager, which is somewhat annoying.

Will flatpak "fix" this, or will it also be released daily?

**Comment by cboxdoerfer** [comment]
The PPA are development packages which are built every time there's a code change, the Flatpak packages will be release builds, so they only get updated when I publish a new release - likely every couple months.

**Comment by Philipp91** [comment]
Excellent. Happy to test when it's available.

Speaking from experience with other flatpak applications: Have you figured out the sandbox? By default (if the application requests it) they can only access the home directory. IIRC there's an entirely different way to access other directories, and it has to be enabled ("connected") manually I believe. This doesn't matter for many applications like a note-taking app (which conceptually just has its own internal storage and nobody cares where that is). But for something like FSearch, file system access is crucial.

**Comment by cboxdoerfer** [comment]
Yes, FSearch needs wide filesystem access to work properly. Flatpak allows you to set `filesystem=host` or `filesystem=home` in the flatpak manifest, which gives the application either access to almost everything or only the home directory. I'll probably use the `host` option by default, which works reasonably well. But or course, if someone really needs to access everything, they should use the "native" package instead of the flatpak version.

**Comment by probonopd** [comment]
AppImages have no enforced sandbox, so that application has full access to the system.

**Comment by Philipp91** [comment]
AFAIU, AppImages are just files without their own distribution system, whereas with Flathub I could get updates automatically.

**Comment by probonopd** [comment]
Check out AppImageUpdate :)

**Comment by cboxdoerfer** [comment]
For everyone interested, it's now available as a Flatpak: https://flathub.org/apps/details/io.github.cboxdoerfer.FSearch

Now I'll have a look at AppImages and Snaps.

**Comment by folknor** [comment]
Another vote for AppImage, which is by far the best solution. I hope flatpak and snap die sooner rather than later, and that ubuntu and redhat get back to deb/rpm instead. But I realise that's not going to happen, and it's unfortunate. I see you provide debs (via ppa), so I'll use those. But if you weren't providing debs, I'd want an AppImage.

AppImage ftw.

**Comment by cboxdoerfer** [comment]
I'm not very familiar with AppImage yet — only did some brief testing with it to get FSearch to run. 

I'll try to provide an AppImage for testing in the next few weeks and if they turn out to work well with FSearch and don't require a lot of work to maintain, then I don't see any reason why we shouldn't have them.

**Comment by phil294** [comment]
Hello,

you said you already did some brief testing, but just in case it's useful to you: I successfully built fsearch as an AppImage with a single command using [linuxdeploy](https://github.com/linuxdeploy/linuxdeploy/) with its [gtk plugin](https://github.com/linuxdeploy/linuxdeploy-plugin-gtk), on a debian 10 docker image (`--privileged` to enable libfuse):

    ./linuxdeploy-x86_64.AppImage --appdir AppDir --output appimage --executable ./fsearch --plugin gtk --desktop-file ./io.github.cboxdoerfer.FSearch.desktop --icon-file ./io.github.cboxdoerfer.FSearch.svg

results in a 28 MiB executable which also runs fine on ArchLinux. Only problem is that the app isn't themed :-/ not sure if that's fixable

**Comment by cboxdoerfer** [comment]
@phil294, thanks, I tried `linuxdeploy` as well and it did work, but unfortunately it's still in an alpha state, so I don't feel comfortable yet to create official builds with it. Do you happen to know if there are any plans for a stable release in the near future?

The `appimage-builder` tool also worked to a degree, but I remember there were some runtime issues when I tried to run the FSearch AppImage on a different distribution (some missing or incompatible library, I forgot which one it was exactly). Not sure if this was because I missed something or because AppImage's aren't truly self-contained.

**Comment by phil294** [comment]
no, they should indeed be truly self-contained. I had the impression that appimage-builder indeed leaves out some dependencies if you haven't installed them but I have not tested it long enough. It seems that linuxdeploy only allows a complete build when all dependencies could be found, a desktop file is present etc., which was reassuring to know when I used it. I have no ideas regarding its release plans, no, sorry, I'm only experimenting myself with a few apps.

**Comment by kendofriendo** [comment]
-1 for Appimage support

**Comment by tbone2k-git** [comment]
> -1 for Appimage support

Why is that? o)

I mean, I could do without AppImages on Linux as well, IF the applications would just run from their own folder and also look in that folder for the belonging resources. Unfortunately most of them don't work like this, but "fsearch" does not need much external resources. So you don't necessarily need an AppImage for it I think, but it's good to have nonetheless.

An AppImage (if you copy it) gives you the option of having multiple fsearch instances with separate database(s) e.g..

You can also just put the "fsearch" executable in its own folder, maybe add an icon and it runs fine from there. It seems so at least, I did not run this on another linux box without "fsearch" being installed by native package manegement as well yet.

For me, the points about using an AppImage are:
- have all application data in one place / folder
- have all dependencies in one place / folder
- be able to copy and store anywhere
- run from anywhere, USB / network etc.. 
- be able to redirect "home" if desired, so it does not mess with your local home folder, it has it's own
- be able to redirect ".config" if desired, so it does not mess with your local .config folder, it has it's own
- redirected .config to a folder next to the executable allows for applications settings to travel along

Wrapping it up:
The application shall be portable and don't require changes or additional packages added to the base system.
An AppImage can give you all that, but it is not required for all applications to be portable (I guess).

I'm kind of new to Linux, but I'm very used to be able to run basically any application directly from its folder, it makes things very easy, especially building new systems, just copy the applications you like over to the new computer, done! Even the settings you used before are already there, nice! o)

For heavy application packages, this often does not work, they spread to much things around unnecessarily, but 90% of applications will start fine from their folder without any prior installation, which I think is a good thing.

This is the bash script I use to run a "portable" fsearch, with database and settings redirected to folders in the application folder of "fsearch-portable", stored somewhere in my application collection folder hierarchy "/bin/file/search/fsearch" or something (to get you an idea). I don't use any system menu to launch applications, to much of a hassle to maintain, I just run things from a well sorted folder structure.

The upper part of the script is some kind of overengineered way of getting the folder the bash script resides in, it follows symlinks and whatnot, you can do that much simpler, the actual part of interest is the lower part, where settings and home folder get redirected to be "next to the application".

I seem to have renamed "fsearch" into "fsearch-portable" here as well, because fsearch does not bring up the correct instance if you have multiple of them running and their executable have the same name. It seems fsearch does not use the executable path to determine the correct instance to bring in to the foreground - just a guess. 

`#!/bin/bash

SOURCE=${BASH_SOURCE[0]};
while [ -L "$SOURCE" ]; do # resolve $SOURCE until the file is no longer a symlink
  DIR=$( cd -P "$( dirname "$SOURCE" )" >/dev/null 2>&1 && pwd );
  SOURCE=$(readlink "$SOURCE");
  # if $SOURCE was a relative symlink, we need to resolve it relative to the
  # path where the symlink file was located
  [[ $SOURCE != /* ]] && SOURCE=$DIR/$SOURCE;
done
SCRIPT_DIR=$( cd -P "$( dirname "$SOURCE" )" >/dev/null 2>&1 && pwd );
SCRIPT_NAME=$(basename -- "$(readlink -f -- "$0")");

pushd "$SCRIPT_DIR"

export HOME="$SCRIPT_DIR/fsearch-portable.home"
export XDG_CONFIG_HOME="$SCRIPT_DIR/fsearch-portable.config"

echo "Running fsearch-portable.."
"$SCRIPT_DIR/fsearch-portable" &

popd
echo "done."`

**Comment by Parsnip** [comment]
Heya friends.
Any updates or news about this maybe hopefully happening at some point?
With the recent bollocks about AUR in my mind I figured I'd see if I can reduce the amount of stuff I have installed from there and fsearch is one of them.

**Comment by grgr4293** [comment]
I agree and would also really like to see an AppImage version if possible.

**Comment by SanmayJoshi** [comment]
For anyone searching for appimage, feel free to get it from https://github.com/SanmayJoshi/fsearch-appimage. Let me know if you face any issues.

**Comment by tbone2k-git** [comment]
> For anyone searching for appimage, feel free to get it from https://github.com/SanmayJoshi/fsearch-appimage. Let me know if you face any issues.

Awesome, thank you! Will try! o)
